### PR TITLE
Add tests for RateLimitTransport

### LINKED ISSUE #53: Add test for rate-limiter
See [here](https://github.com/src-d/ghsync/pull/51#issuecomment-506749452).

An alternative could be to run a test server that with some condition answers with the needed headers.

### PR DESCRIPTION
**Pull request by smacker**
Fix: #53

Coverage:
Screenshot 2019-08-06 at 18 10 23

### CODE REVIEW

File: utils/rate_test.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+		c.getSuccess()
+		c.getSuccess()
+	})
+	assert.True(spent < time.Millisecond)
+
+	// should sleep before success for more than a second
+	spent = measure(func() {
+		c.getSuccess()
+	})
+	assert.True(spent > 100*time.Millisecond)
```
Comment (se7entyse7en) [inline utils/rate_test.go]: Shouldn't this be nearly `2*time.Second`? Why only `> 100*time.Millisecond`?

File: utils/rate_test.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+// emulates abuse error
+type abuseTransport struct {
+	RetryAfter int // number of seconds
+
+	returnSuccess bool
+}
+
+func (t *abuseTransport) RoundTrip(*http.Request) (*http.Response, error) {
+	if t.returnSuccess {
```
Comment (se7entyse7en) [inline utils/rate_test.go]: Doesn't seem to be ever used.

File: utils/rate_test.go (hunk 1 above)
Comment (se7entyse7en) [reply utils/rate_test.go]: It's not very different as the real comparison is _immediate_ vs _non-immediate_, but still, maybe it's more informative writing for example `> 1*time.Second` as you did for example [here](https://github.com/src-d/ghsync/pull/61/files#diff-a95f4af213b65e756e16458237fcf615R84).

**Review by se7entyse7en — CHANGES_REQUESTED** [review]

**Comment by se7entyse7en** [comment]
Cool BTW, and great coverage 👏.

File: utils/rate_test.go (hunk 2 above)
Comment (smacker) [reply utils/rate_test.go]: it does. If `abuseTransport` never returns success `RateLimitTransport` would go into infinite loop.

File: utils/rate_test.go (hunk 1 above)
Comment (smacker) [reply utils/rate_test.go]: thanks! Fixed. I was playing with smaller numbers and forgot to revert it here. (smaller timeouts make tests unstable sometimes)

File: utils/rate_test.go (hunk 2 above)
Comment (se7entyse7en) [reply utils/rate_test.go]: Oh ok, I got it now. Thanks!

**Review by se7entyse7en — APPROVED** [review]

File: utils/rate_test.go (hunk 1 above)
Comment (se7entyse7en) [reply utils/rate_test.go]: > smaller timeouts make tests unstable sometimes

Yeah, I came across this once and read something about how the minimum amount of sleep that an OS is able to guarantee.

Comment (smacker) [reply utils/rate_test.go]: OS sleep interval is much smaller, ~15ms. In this case, the go code was executing too slow when my laptop was under heavy load and it was failing with ~100ms timeouts.

**Review by carlosms — APPROVED** [review]

**Pull request merged**